File: rust/0290-word-pattern.rs
impl Solution {
    pub fn word_pattern(pattern: String, s: String) -> bool {
        use std::collections::{HashMap, HashSet};

        if pattern.len() != s.split_whitespace().count() {
            return false;
        }

        let mut char_to_word = HashMap::new();
        let mut word_to_char = HashMap::new();

        for (c, word) in pattern.chars().zip(s.split_ascii_whitespace()) {
            if let Some(w) = char_to_word.insert(c, word) {
                if w != word {
                    return false;
                }
            }

            if let Some(w) = word_to_char.insert(word, c) {
                if w != c {
                    return false;
                }
            }
        }
        true
    }
}

File: csharp/0290-word-pattern.cs
public class Solution {
    public bool WordPattern(string pattern, string s) {
        var words = s.Split(' ');
        if (words.Length != pattern.Length) return false;

        Dictionary<char,string> charToWord = new Dictionary<char, string>();
        Dictionary<string, char> wordToChar = new Dictionary<string, char>();

        for (int i = 0; i < pattern.Length; i++) {
            char c = pattern[i];
            string w = words[i];
            if (charToWord.ContainsKey(c) && (charToWord[c] != w)) {
                return false;
            }
            if(wordToChar.ContainsKey(w) && (wordToChar[w] != c)) {
                return false;
            }
            charToWord[c] = w;
            wordToChar[w] = c;
        }
        return true;
    }
}